Action creator helpers

jstorm31 opened the pull request:
I added action creator helpers for creating generic request actions and/or types to reduce the amount of boilerplate code. documentation in `docs/actions/request.md`
-  `reqeustActions`
-  `requestActionsTypes`
- `createRequestActionsTypes`

I also fixed`jest` configuration.

Closes #23 
Closes #24

jstorm31 commented:
Please let me know if you agree with the API design, naming, etc.

cermakjiri commented on docs/actions/request.md:
````diff
@@ -0,0 +1,101 @@
+});
+```
+
+## <a name="createRequestActionsTypes"></a> `createRequestActionsTypes(config: Object) => Function`
````
What about to have just the first `requestActions` utility? I can use it with without linking the logic for creating action types and that's good. You can choose how do you want to create the action types and then pass them to the `requestActions` utility.  The least utils, the better.

jstorm31 replied:
Just to clarify, do you mean to only have `requestActions` utility and to always delegate the responsibility of creating types?

Personally, I find `createRequestActionsTypes` more useful exactly for the reduced amount of work I have to do manually. The latter one is only a few more lines of code and for convenience I think it's fine to provide both.

cermakjiri replied:
Yes, exactly. I'd rather have simplier non-dependent blocks - utilts, that you can put together than one monolithic utilities with complex configuration. Also sometime you will want to use the two blocks in separated ways. But since it's decision that will apply to all of us, let's discuss it tomorrow at the frontend meeting.

jstorm31 replied:
Good, I removed everything but `apiRequestActions`. Types will have to be created manually.

jstorm31 commented on docs/actions/request.md:
```diff
@@ -0,0 +1,101 @@
+
+/*
+{
+    fetchArticlesRequest: params => ({
+        type: 'articles/FETCH_ARTICLES_REQUEST',
+        payload: { ...params }
+    }),
+    fetchArticlesSuccess: (payload, meta) => ({
+        type: 'articles/FETCH_ARTICLES_SUCCESS',
+        meta,
+        payload
+    }),
+    fetchArticlesFailure: error => ({
+        type: 'articles/FETCH_ARTICLES_FAILURE',
+        error
+    }),
+    fetchArticlesCancel: () => ({
+        type: 'articles/FETCH_ARTICLES_CANCEL'
+    }),
+    fetchArticlesReset: () => ({
+        type: 'articles/FETCH_ARTICLES_RESET'
+    })
+}
```
1. Good point! 👍  I only thought about requests for all entities, not a detail. I'll think about a solution.
2. Yep, `apiRequestActions` would definitely be better. I also though about `asyncRequests` to keep the convention with `asyncTypes`, but I don't understand what "async" exactly means here.

cermakjiri replied:
The `api` prefix is better, the `async` prefix is rather older name convention. Maybe it'd make sense to rename the asyncTypes to apiRequestTypes 🤔 Anyway that's not a subject of this PR.

jstorm31 replied:
1. @cermakjiri  Couldn't you use the `params` parameter of the request action creator for the ID?
```js
const { types, actions } = apiRequestActionsTypes({
    modulePrefix: 'article',
    typePrefix: 'FETCH_ARTICLE_',
});

// Calling the action to fetch an article by ID
const articleId = 45
actions.fetchArticleRequest({ articleId })
```
2. I renamed it to `apiRequestActions` , etc.

jstorm31 replied:
I unified the naming to `apiRequest[Type|Actions]` therefore this will be a breaking change.

cermakjiri replied:
> 1. @cermakjiri  Couldn't you use the `params` parameter of the request action creator for the ID?
> 
> ```js
> const { types, actions } = apiRequestActionsTypes({
>     modulePrefix: 'article',
>     typePrefix: 'FETCH_ARTICLE_',
> });
> 
> // Calling the action to fetch an article by ID
> const articleId = 45
> actions.fetchArticleRequest({ articleId })
> ```
> 
> 1. I renamed it to `apiRequestActions` , etc.

Unfortunately, that's not a good solution that will be compatible with api reducers. Each such an action has to include item ID.

```js
fetchUserRequest(userId, params)
fetchUserSuccess(userId, payload, meta)
fetchUserCancel(userId)
...

// and also
fetchUsersRequest(groupId, params)
...
```

I guess the current api could be only extended by `options` object as 2nd argument:
```js
const types =  apiRequestType({
   typePrefix: 'FETCH_USER_'
});

const actions = apiRequestActions(types, {
   // I don't like this but it'd be easy to implement :D 
   firstArgumentIsId: true
});

// usage:
actions.fetchUserRequest(userId)
actions.fetchUserSuccess(userId, payload)
```

I am not sure if there should some mapping function, custom args parser function or just boolean option.

jstorm31 replied:
I don't see anything wrong with your proposed solution and I implemented it. I named the parameter `isDetailRequest` though. As an alternative, I can think of accepting any number of parameters by the action creator `(...params) => ({ type: ..., ...params })`, but I think your solution is better. Check out the new changes.

cermakjiri replied:
The new changes look good. I'd suggest to published it with the `alpha` tag and if we won't be happy with current solution, we can then fallback to yours.

jstorm31 commented on docs/actions/request.md:
````diff
@@ -0,0 +1,101 @@
+```js
+import { requestActionsTypes } from '@ackee/redux-utils';
+
+const { types, actions } = requestActionsTypes({
````
Good tip 👍 . But this code has been removed since you didn't agree with keeping it.

jstorm31 commented on src/utilities/strictObjectAccess/strictObjectAccess.test.js:
```diff
@@ -1,7 +1,7 @@
 
-it('Should throw an error if the object property value is undefined', () => {
-    const objectWithStrictAccess = strictObjectAccess({});
+it.only('Should throw an error if the object property value is undefined', () => {
```
Good 👁 . Thanks!

jstorm31 commented on src/actions/request/requestActionsTypes.js:
```diff
@@ -0,0 +1,28 @@
+
+/**
+ * Generates a factory to create Redux actions and types with predefined module prefix and default types
+ * @param {Object} config { modulePrefix: String, defaultTypes: Array<Sting> }
```
Thanks 👍

jstorm31 commented:
I updated the PR and changed a couple of things:

- Removed `apiRequestActionsTypes` and `createApiRequestActionsTypes` since you didn't think they are necessary
- Renamed `asyncType` to `apiRequestType` (this is a **breaking** change)
- Added tests for `apiRequestType` helper
- Updated docs

Please review the changes and let me know if you have any further comments.

jstorm31 commented on src/actions/apiRequest/templates.js:
```diff
@@ -0,0 +1,12 @@
+        type,
+        payload: {
+            ...params,
+        },
```
Yes 👍🏻
